Source/WebCore/dom/ContentVisibilityDocumentState.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2023 Igalia S.L. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "IntersectionObserver.h"
+
+namespace WebCore {
+
+class Document;
+class Element;
+
+class ContentVisibilityDocumentState {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static void observe(Element&);
+    static void unobserve(Element&, Document&);
+
+    void updateContentRelevancyStatusForScrollIfNeeded(const Element& scrollAnchor);
+
+    bool hasObservationTargets() const { return m_observer && m_observer->hasObservationTargets(); }
+
+    void updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancyStatus>&);
+
+    void updateOnScreenObservationTarget(const Element&, bool);
+
+private:
+    IntersectionObserver* intersectionObserver(Document&);
+
+    RefPtr<IntersectionObserver> m_observer;
+
+    Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> m_onScreenObservationTargets;
```

Comment:
I would suggest using a `WeakHashSet` here, since it's just a Set that doesn't need to be ordered.

Reply:
Done.